### ISSUE TITLE
The distance function does not reach the index

### ISSUE THREAD

**Issue by bingo919** [neo4j/neo4j#12215]
Neo4j version:3.4.1
Operating system: centos6.5

eg1:**When there are more than five pieces of data, you can't go to the index with distance**

```
match(n:City) detach delete n;                                                                      
drop CONSTRAINT ON (cc:City) ASSERT cc.location IS UNIQUE   ;                                       
                                                                                                    
                                                                                                    
CREATE (bruges2:City {name:'bruges', location:point({ srid:4326, srid:4326, x: 101.585997,y: 30.342134})});   
CREATE (bruges:City {name:'bruges', location:point({ srid:4326, x: 102.585997,y: 31.342134})});    
CREATE (brussels:City {name:'brussels', location:point({ srid:4326, x: 103.585997,y: 32.342134})});
CREATE (paris:City {name:'paris', location:point({ srid:4326, x: 104.585997,y: 33.342134})});      
CREATE (paris1:City {name:'paris1', location:point({ srid:4326, x: 105.585997,y: 34.342134})});    
CREATE (paris2:City {name:'paris2', location:point({ srid:4326, x: 106.585997,y: 35.342134})});    
     
create CONSTRAINT ON (cc:City) ASSERT cc.location IS UNIQUE   ;                                     
                                                                                               
explain                                                                                             
match(c:City)                                                                                       
WHERE distance(c.location, point({ x:104.585997, y:  30.342134 }))< 200                   
RETURN c
```;     


+-------------------------------------------------------------------------+
| Plan      | Statement   | Version      | Planner | Runtime       | Time |
+-------------------------------------------------------------------------+
| "EXPLAIN" | "READ_ONLY" | "CYPHER 3.4" | "COST"  | "INTERPRETED" | 5    |
+-------------------------------------------------------------------------+

+------------------+----------------+-------------+-----------------------------------------------------------------------------------------+
| Operator         | Estimated Rows | Identifiers | Other                                                                                   |
+------------------+----------------+-------------+-----------------------------------------------------------------------------------------+
| +ProduceResults  |              1 | c           | 3.4; 3.4                                                                                |
| |                +----------------+-------------+-----------------------------------------------------------------------------------------+
| +Filter          |              1 | c           | distance(c.location, point({x: $`  AUTODOUBLE0`, y: $`  AUTODOUBLE1`})) < $`  AUTOINT2` |
| |                +----------------+-------------+-----------------------------------------------------------------------------------------+
| +NodeByLabelScan |              6 | c           | :City                                                                                   |
+------------------+----------------+-------------+-----------------------------------------------------------------------------------------+

  ```                                      
eg2:  When the data is less than 5, you can walk to the index using distance
                                                                                             
``` 
match(n:City) detach delete n;                                                                      
drop CONSTRAINT ON (cc:City) ASSERT cc.location IS UNIQUE   ;                                       
                                                                                                    
                                                                                                    
CREATE (bruges2:City {name:'bruges', location:point({ srid:4326, srid:4326, x: 101.585997,y: 30.342134})});   
CREATE (bruges:City {name:'bruges', location:point({ srid:4326, x: 102.585997,y: 31.342134})});    
CREATE (brussels:City {name:'brussels', location:point({ srid:4326, x: 103.585997,y: 32.342134})});
CREATE (paris:City {name:'paris', location:point({ srid:4326, x: 104.585997,y: 33.342134})});      
CREATE (paris1:City {name:'paris1', location:point({ srid:4326, x: 105.585997,y: 34.342134})});    
     
create CONSTRAINT ON (cc:City) ASSERT cc.location IS UNIQUE   ;                                     
                                                                                               
explain                                                                                             
match(c:City)                                                                                       
WHERE distance(c.location, point({ x:104.585997, y:  30.342134 }))< 200                   
RETURN c;     
  

+-------------------------------------------------------------------------+
| Plan      | Statement   | Version      | Planner | Runtime       | Time |
+-------------------------------------------------------------------------+
| "EXPLAIN" | "READ_ONLY" | "CYPHER 3.4" | "COST"  | "INTERPRETED" | 13   |
+-------------------------------------------------------------------------+

+-----------------------------+----------------+-------------+-----------------------------------------------------------------------------------------+
| Operator                    | Estimated Rows | Identifiers | Other                                                                                   |
+-----------------------------+----------------+-------------+-----------------------------------------------------------------------------------------+
| +ProduceResults             |              1 | c           | 3.4; 3.4                                                                                |
| |                           +----------------+-------------+-----------------------------------------------------------------------------------------+
| +Filter                     |              1 | c           | distance(c.location, point({x: $`  AUTODOUBLE0`, y: $`  AUTODOUBLE1`})) < $`  AUTOINT2` |
| |                           +----------------+-------------+-----------------------------------------------------------------------------------------+
| +NodeUniqueIndexSeekByRange |              5 | c           | :City(location) WHERE distance(_,point(x,y)) < Parameter(  AUTOINT2,Integer)            |
+-----------------------------+----------------+-------------+-----------------------------------------------------------------------------------------+

**Comment by craigtaverner** [comment]
With only 5 or 6 nodes both an index or a label scan are probably about the same cost, so it is possible that the planner is not able to tell the difference and can choose either. However, if we see a label scan when there are hundred or thousands of nodes, then we would see this as a bug. Can you confirm the behaviour with hundred or thousands of points?

**Comment by bingo919** [comment]
when  there are hundred or thousands of nodes ，the  bug still exists.
Because I have a lot of data in my production system, I can't reach the index with distance function.
thank you @craigtaverner

**Comment by craigtaverner** [comment]
Thanks for the report @bingo919 , we'll start investigating internally. Can you verify if this same behaviour is happening if you do not have the unique constraint, but only a normal index on `:City(location)`?

**Comment by bingo919** [comment]
```
match(n:City) detach delete n;                                                                      
drop CONSTRAINT ON (cc:City) ASSERT cc.location IS UNIQUE   ;                                       
                                                                                                    
                                                                                                    
CREATE (bruges2:City {name:'bruges', location:point({ srid:4326, srid:4326, x: 101.585997,y: 30.342134})});   
CREATE (bruges:City {name:'bruges', location:point({ srid:4326, x: 102.585997,y: 31.342134})});    
CREATE (brussels:City {name:'brussels', location:point({ srid:4326, x: 103.585997,y: 32.342134})});
CREATE (paris:City {name:'paris', location:point({ srid:4326, x: 104.585997,y: 33.342134})});      
CREATE (paris1:City {name:'paris1', location:point({ srid:4326, x: 105.585997,y: 34.342134})});    
CREATE (paris2:City {name:'paris2', location:point({ srid:4326, x: 106.581997,y: 35.342134})});    
CREATE (paris3:City {name:'paris3', location:point({ srid:4326, x: 106.582997,y: 36.342134})});    
CREATE (paris4:City {name:'paris4', location:point({ srid:4326, x: 106.583997,y: 37.342134})});    
CREATE (paris5:City {name:'paris5', location:point({ srid:4326, x: 106.584997,y: 38.342134})});    
CREATE (paris6:City {name:'paris6', location:point({ srid:4326, x: 106.585997,y: 39.342134})});    
CREATE (paris7:City {name:'paris7', location:point({ srid:4326, x: 106.586997,y: 35.242134})});    


  CREATE index ON :City(location);

                                                                                               
explain                                                                                             
match(c:City)                                                                                       
WHERE distance(c.location, point({ x:104.585997, y:  30.342134 }))< 200                   
RETURN c

+-------------------------------------------------------------------------+
| Plan      | Statement   | Version      | Planner | Runtime       | Time |
+-------------------------------------------------------------------------+
| "EXPLAIN" | "READ_ONLY" | "CYPHER 3.4" | "COST"  | "INTERPRETED" | 9    |
+-------------------------------------------------------------------------+

+-----------------------+----------------+-------------+-----------------------------------------------------------------------------------------+
| Operator              | Estimated Rows | Identifiers | Other                                                                                   |
+-----------------------+----------------+-------------+-----------------------------------------------------------------------------------------+
| +ProduceResults       |              1 | c           | 3.4; 3.4                                                                                |
| |                     +----------------+-------------+-----------------------------------------------------------------------------------------+
| +Filter               |              1 | c           | distance(c.location, point({x: $`  AUTODOUBLE0`, y: $`  AUTODOUBLE1`})) < $`  AUTOINT2` |
| |                     +----------------+-------------+-----------------------------------------------------------------------------------------+
| +NodeIndexSeekByRange |              1 | c           | :City(location) WHERE distance(_,point(x,y)) < Parameter(  AUTOINT2,Integer)            |
+-----------------------+----------------+-------------+-----------------------------------------------------------------------------------------+

0 rows available after 9 ms, consumed after another 0 ms

```

###   **CREATE index ON :City(location);**

### **When I build a normal index on :City(location) , I can go to the index.**

**Comment by Lojjs** [comment]
Hi @bingo919 . I'm currently looking into you bug report. I'm not sure this is the entire problem, but I just wanted you to know that your distance query `WHERE distance(c.location, point({ x:104.585997, y:  30.342134 }))< 200` will always produce null since this point will be in the cartesian coordinate system by default while your other points are in WGS-84. For more info, see considerations under 3.4.10.2 in our documentation: https://neo4j.com/docs/developer-manual/3.4/cypher/functions/spatial/

Best regards Louise, Neo4j Cypher team

**Comment by bingo919** [comment]
Thank you for reminding me @Lojjs 
but I still can't get to the index with  points are in WGS-84

```
match(n:City) detach delete n;                                                                      
drop CONSTRAINT ON (cc:City) ASSERT cc.location IS UNIQUE   ;                                       
                                                                                                    
                                                                                                    
CREATE (bruges2:City {name:'bruges', location:point({ srid:4326, srid:4326, x: 101.585997,y: 30.342134})});   
CREATE (bruges:City {name:'bruges', location:point({ srid:4326, x: 102.585997,y: 31.342134})});    
CREATE (brussels:City {name:'brussels', location:point({ srid:4326, x: 103.585997,y: 32.342134})});
CREATE (paris:City {name:'paris', location:point({ srid:4326, x: 104.585997,y: 33.342134})});      
CREATE (paris1:City {name:'paris1', location:point({ srid:4326, x: 105.585997,y: 34.342134})});    
CREATE (paris2:City {name:'paris2', location:point({ srid:4326, x: 106.585997,y: 35.342134})});    
     
create CONSTRAINT ON (cc:City) ASSERT cc.location IS UNIQUE   ;    

neo4j> explain                                                                                             
       match(c:City)                                                                                       
       WHERE distance(c.location, point({ srid:4326, x:104.585997, y:  30.342134 }))< 200                   
       RETURN c;
+-------------------------------------------------------------------------+
| Plan      | Statement   | Version      | Planner | Runtime       | Time |
+-------------------------------------------------------------------------+
| "EXPLAIN" | "READ_ONLY" | "CYPHER 3.4" | "COST"  | "INTERPRETED" | 11   |
+-------------------------------------------------------------------------+

+------------------+----------------+-------------+--------------------------------------------------------------------------------------------------------------+
| Operator         | Estimated Rows | Identifiers | Other                                                                                                        |
+------------------+----------------+-------------+--------------------------------------------------------------------------------------------------------------+
| +ProduceResults  |              1 | c           | 3.4; 3.4                                                                                                     |
| |                +----------------+-------------+--------------------------------------------------------------------------------------------------------------+
| +Filter          |              1 | c           | distance(c.location, point({srid: $`  AUTOINT0`, x: $`  AUTODOUBLE1`, y: $`  AUTODOUBLE2`})) < $`  AUTOINT3` |
| |                +----------------+-------------+--------------------------------------------------------------------------------------------------------------+
| +NodeByLabelScan |              6 | c           | :City                                                                                                        |
+------------------+----------------+-------------+--------------------------------------------------------------------------------------------------------------+

```
eg:  distance(c.location, point({ srid:4326, x:104.585997, y:  30.342134 }))< 200                   
thanks

**Comment by Lojjs** [comment]
@bingo919 After some further investigation, we found out that there was indeed a bug in planning of NodeUniqueIndexSeekByRange (the uniqueness constraint counterpart to NodeIndexSeekByRange) in the case of spatial points. A fix will be included in the next maintenance releases of 3.4 and 3.5 (3.4.15 and 3.5.7).
